fenicsxprecice/adapter_core.py

```diff
@@ -122,58 +155,210 @@ def convert_fenicsx_to_precice(fenicsx_function, local_coords):
+        # to avoid UnboundLocalError, interpolation_coordinates is an array to which x is appended to
+        interpolation_coordinates.append(np.transpose(copy.deepcopy(x)))
+        # directly round and make each rounded coordinate unique to keep the code clean and to reduce memory consumption
+        interpolation_coordinates[-1] = round_unique_coordinates(interpolation_coordinates[-1])
+
+        return np.zeros((vec_len, x.shape[1]))
+
+    # determine process local cells that are on the coupling boundary
+    dofs_coupling = fem.locate_dofs_geometrical(function_space, coupling_boundary)
+    dofs_coupling_coordinates = function_space.tabulate_dof_coordinates()[dofs_coupling]
+    bb_tree = geometry.bb_tree(domain, domain.geometry.dim)
+    cell_candidates_local = geometry.compute_collisions_points(bb_tree, dofs_coupling_coordinates)
+    cell_candidates_local = geometry.compute_colliding_cells(domain, cell_candidates_local, dofs_coupling_coordinates)
+    # the cell candidates with local cell ids
+    cell_candidates_local = np.unique(cell_candidates_local.array)
+
+    # determine process-owned cells
+    index_map = domain.topology.index_map(domain.topology.dim)
+    # range of owned cells
+    owned_cell_ids = index_map.local_range
+    owned_cell_ids = np.arange(owned_cell_ids[0], owned_cell_ids[1])
+    # map local cell ids to global cell ids to determine ghost cells
+    cell_candidates_global = index_map.local_to_global(cell_candidates_local)
+    # cells to be interpolated over must be owned! Find intersection of cell candidates and owned cells
+    owned_and_candidate_cells_global = np.intersect1d(owned_cell_ids, cell_candidates_global)
+    # map back to local cell indexing
+    owned_and_candidate_cells_local = index_map.global_to_local(owned_and_candidate_cells_global)
+
+    # get all interpolation points of owned domain
+    query_function.interpolate(query_coordinates, owned_and_candidate_cells_local)
+    # convert the numpy array to a set of tuples to allow set operations
+    interpolation_coordinates = set(map(tuple, interpolation_coordinates[0]))
+
+    # to avoid creating an ill-posed mapping problem for preCICE (especially RBF mapping), equal coordinates used by multiple MPI ranks
+    # must be determined and removed by all but one MPI rank.
+
+    # rule: if rankA and rankB use coordinate x and rankA < rankB, rankA keeps
+    # x and rankB discards the coordinate, else rankB keeps x
+
+    # send interpolation points of complete boundary (interpolation_coordinates[0]) to higher ranks
+    # OR receive interpolation points that potentially need to be filtered out from lower ranks
+
+    # TODO add persistent communication for performance increase?
+
+    duplicate_coordinates_per_rank = {}  # save the filtered coordinates per rank to reduce communication volume
+
+    for source_rank in range(comm_rank):
+        interpolation_coordinates_from_source = comm.recv(source=source_rank)  # recv a set of tuples
```

Comment:
Does this cover the scenario when a rank needs to send as well as receive? Or would this never be the case in the algorithm?